activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb

```diff
@@ -48,26 +48,39 @@ def perform_query(raw_connection, sql, binds, type_casted_binds, prepare:, notif
             # made since we established the connection
             raw_connection.query_options[:database_timezone] = default_timezone
 
-            result = if prepare
-              stmt = @statements[sql] ||= raw_connection.prepare(sql)
+            result = if !prepared_statements || binds.nil? || binds.empty?
```

Comment:
This shouldn't be checking `prepared_statements`

Reply:
I was going based off this:
https://github.com/rails/rails/blob/7-1-stable/activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb#L22

Which calls the old `without_prepared_statements` here:
https://github.com/rails/rails/blob/d5e70917426d843a7b6139b75cc0c3146d69cbd0/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L1174-L1176

Reply:
It doesn't make sense to check it here. If we get here with binds, we need to do the prepare+execute: dropping the binds on the floor and sending a placeholder-ed string into `query` is never going to work.